### DISCUSSION TITLE
Reactive getter

### DISCUSSION THREAD

**Discussion by anubra266** [pmndrs/zustand#1519] [Q&A]
Say I have a state in my store called `list`, it's an array of objects. Each object has an Id.

I want a `getItem` method that does something like 
```js
getItemById: get().list.find(item => item.id === itemId)
```

And I need it to be reactive, so If I use this to get an item in my `<ItemDetails />` component, mutating the item in `list` would reflect immediately.

Right now all I could do was use a hook
![code](https://user-images.githubusercontent.com/30869823/211316040-8a134782-4040-4747-b5bc-9897899aa56a.png)

**Comment by dai-shi** [comment]
Your `useWatchListItem` isn't ideal. You can return derived value from the selector:
```js
export const useWatchListItem = (id) => {
  return useStore((state) => state.watchList.find((li) => li.id.toString() === id.toString()))
}
```
That's the recommended way.

**Reply by anubra266** [reply]
Thanks. So there's no way to create derived state in store

**Reply by dai-shi** [reply]
There are some third party libraries for such use cases. But, yeah, using the hook is the primary way.

**Reply by shinyruo-nmsl** [reply]
> There are some third party libraries for such use cases. But, yeah, using the hook is the primary way.有一些第三方库可以解决此类问题。不过，使用钩子仍然是主要方法。

in my case, i need to derived state in store. otherwise, I would need to write a lot of repetitive code. Can you introduce some useful third-party libraries

**Reply by dai-shi** [reply]
https://github.com/zustandjs/derive-zustand is the one I tried.
However, if you just want to avoid repetitive code, I guess defining a custom hook is easier and recommended.

**Reply by SetsuikiHyoryu** [reply]
> Thanks. So there's no way to create derived state in store

> in my case, i need to derived state in store. otherwise, I would need to write a lot of repetitive code. Can you introduce some useful third-party libraries

My code looks like this. `employeeName` is computed from `employeeInfo`, and it is reactive.

```TypeScript
import type { EmployeeInfo } from '@/features/authorization'
import { Option } from '@/shared/utils/rust'
import { create } from 'zustand'

type EmployeeStore = {
  // State
  token: string
  // Action
  setToken: (token: string) => void

  employeeInfo: Option<EmployeeInfo>
  setEmployeeInfo: (info: Option<EmployeeInfo>) => void

  // Getter
  employeeName: () => Option<string>
}

export const useEmployeeStore = create<EmployeeStore>()((set, get) => ({
  token: '',
  setToken: token => set(() => ({ token })),

  employeeInfo: Option.none(),
  setEmployeeInfo: info => set(() => ({ employeeInfo: info })),

  employeeName: () => {
    const { employeeInfo: info } = get()
    return info.and_then(item => item.name)
  }
}))
```

**Reply by shinyruo-nmsl** [reply]
> https://github.com/zustandjs/derive-zustand is the one I tried.https://github.com/zustandjs/derive-zustand 是我尝试过的。 However, if you just want to avoid repetitive code, I guess defining a custom hook is easier and recommended.但是，如果您只是想避免重复代码，我想定义自定义钩子会更容易，也更值得推荐。

in some case, i need to use derived state in store, custom hook can't be used in store, it may be a problem, just like this:
```typescript
export function createSearchParamsSlice(): StateCreator<SearchParamsSlice, [], [], SearchParamsSlice> {
  return (set, get) => ({
     // ...
    getOutboundDate: () => {
      return get().getCurQuery()?.outwardDate || format(new Date(), 'yyyy-MM-dd');
    },
    getPreSaleStartDate: () => {
      const startDate = get().searchParams?.preSaleInfo.startDate;
      if (!startDate) {
        return get().getOutboundDate();
      }

      return startDate;
    },
    getDisabledInboundDates: () => {
      const { getOutboundDate, getPreSaleStartDate } = get(); // i need use derived state in store : getOutboundDate/getPreSaleStartDate
      const outboundDate = getOutboundDate();
      const startDate = getPreSaleStartDate();
      const endDate = format(addDays(new Date(getCompatibleDate(outboundDate)), 0), 'yyyy-MM-dd');
      return getDatesInRanges(startDate, endDate);
    },
  });
}
```
